tests/devices/qubit/test_preprocess.py

```diff
@@ -0,0 +1,377 @@
+            else:
+                assert qml.equal(t.measurements[0], measurements[1])
+
+        # Replace with Python Device once added
+        dev = qml.device("default.qubit", wires=2)
+        expected = qml.execute(tapes, dev)
+        res = batch_fn(dev.batch_execute(res_tapes))
+        assert np.allclose(res, expected)
+
+    def test_preprocess_expand(self):
+        """Test that preprocess returns the correct tapes when expansion is needed."""
+        ops = [qml.Hadamard(0), NoMatOp(1), qml.RZ(0.123, wires=1)]
+        measurements = [[qml.expval(qml.PauliZ(0))], [qml.expval(qml.PauliX(1))]]
+        tapes = [
+            QuantumScript(ops=ops, measurements=measurements[0]),
+            QuantumScript(ops=ops, measurements=measurements[1]),
+        ]
+
+        res_tapes, batch_fn = preprocess(tapes)
+        expected = [qml.Hadamard(0), qml.PauliX(1), qml.PauliY(1), qml.RZ(0.123, wires=1)]
+
+        assert len(res_tapes) == 2
+        for i, t in enumerate(res_tapes):
+            for op, exp in zip(t.circuit, expected + measurements[i]):
+                assert qml.equal(op, exp)
+
+        # Replace with Python Device once added
+        dev = qml.device("default.qubit", wires=2)
+        expected = qml.execute(tapes, dev)
+        res = batch_fn(dev.batch_execute(res_tapes))
+        assert np.allclose(res, expected)
+
+    def test_preprocess_split_and_expand(self):
+        """Test that preprocess returns the correct tapes when splitting and expanding
+        is needed."""
+        ops = [qml.Hadamard(0), NoMatOp(1), qml.RX([np.pi, np.pi / 2], wires=1)]
+        # Need to specify grouping type to transform tape
+        measurements = [qml.expval(qml.PauliX(0)), qml.expval(qml.PauliZ(1))]
+        tapes = [
+            QuantumScript(ops=ops, measurements=[measurements[0]]),
+            QuantumScript(ops=ops, measurements=[measurements[1]]),
+        ]
+
+        res_tapes, batch_fn = preprocess(tapes)
+        expected_ops = [
+            [qml.Hadamard(0), qml.PauliX(1), qml.PauliY(1), qml.RX(np.pi, wires=1)],
+            [qml.Hadamard(0), qml.PauliX(1), qml.PauliY(1), qml.RX(np.pi / 2, wires=1)],
+        ]
+
+        assert len(res_tapes) == 4
+        for i, t in enumerate(res_tapes):
+            for op, expected_op in zip(t.operations, expected_ops[i % 2]):
+                assert qml.equal(op, expected_op)
+            assert len(t.measurements) == 1
+            if i < 2:
+                assert qml.equal(t.measurements[0], measurements[0])
+            else:
+                assert qml.equal(t.measurements[0], measurements[1])
+
+        # Replace with Python Device once added
+        dev = qml.device("default.qubit", wires=2)
+        expected = qml.execute(tapes, dev)
+        res = batch_fn(dev.batch_execute(res_tapes))
+        # Need to check each row individually because the dimensions of ``res`` and
+        # ``expected`` are different
```

Comment:
what are the different dimensions? will this still be the case with the Python Device?

Reply:
`qml.enable_return` probably.  We are configuring the new device to only work under the new return types specification.

Reply:
Yep! It was because of the new return types.